### PR TITLE
add SettingsCheckbox components

### PR DESCRIPTION
**Pull request by GretaD**

### CODE REVIEW

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+		}
+	},
+	beforeCreate: function() {
+		this.uuid = uuid.toString()
+		uuid += 1
+	}
+}
+</script>
+
+<style scoped>
```
Comment (juliushaertl) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]:
```suggestion
<style scoped>
p {
	margin-bottom: 15px;
}
```

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 2]
```diff
@@ -0,0 +1,81 @@
+<!--
+  - @copyright Copyright (c) 2019 Julius Härtl <jus@bitgrid.net>
+  -
+  - @author Julius Härtl <jus@bitgrid.net>
```
Comment (georgehrke) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: @GretaD Add your own copyright, not Julius' :P

Comment (GretaD) [reply src/components/SettingsCheckbox/SettingsCheckbox.vue]: I added myself only on the documents that i edited myself. this one I got it from richdocuments done by Julius and didnt really change anything, so I thought would be fair this way :)

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  - but WITHOUT ANY WARRANTY; without even the implied warranty of
+  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  - GNU Affero General Public License for more details.
+  -
+  - You should have received a copy of the GNU Affero General Public License
+  - along with this program. If not, see <http://www.gnu.org/licenses/>.
+  -
+  -->
+
+<template>
```
Comment (juliushaertl) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: A <docs> section would be nice :)

Comment (juliushaertl) [reply src/components/SettingsCheckbox/SettingsCheckbox.vue]: A sorry, my typing skills failed me. I meant a `<docs>` section, so we have an example how to use this component and how it looks like in the documentation. :see_no_evil:

Comment (juliushaertl) [reply src/components/SettingsCheckbox/SettingsCheckbox.vue]: See https://github.com/nextcloud/nextcloud-vue/pull/671/files#diff-e00e2c81e3689faed3469a4bbfd5964fR23-R64 for an example

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  - along with this program. If not, see <http://www.gnu.org/licenses/>.
+  -
+  -->
+<docs>
+## One checkbox
+
+```vue
+	<template>
+		<div>
+			<b-form-checkbox
```
Comment (juliushaertl) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: What is b-form-checkbox?

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	</template>
+	<script>
+		export default {
+			data() {
+				return {
+					status: {{ status }}
+				}
+			}
+		}
+	</script>
```
Comment (juliushaertl) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: Missing end of the vue code block 
` ``` `

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+			data() {
+				return {
+					selected: [], // Must be an array reference
+					options: [
+						{ text: '', value: '' },
+					]
+				}
+			}
+		}
+	</script>
```
Comment (juliushaertl) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: Also missing the end of the code block

File: src/components/SettingsCheckbox/SettingsCheckbox.vue (hunk 4 above)
Comment (GretaD) [reply src/components/SettingsCheckbox/SettingsCheckbox.vue]: For cross browser consistency,  <b-form-checkbox> use Bootstrap's custom checkbox input to replace the browser default checkbox input. 
Should I keep it simple: 
<input
  type="checkbox"> ?

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	}
+</script>
+```
+
+</docs>
+
+<template>
+	<div class="section">
+		<p>
+			<input :id="id" type="checkbox" class="checkbox"
```
Comment (georgehrke) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: Please put each property into a new line for readability, thx!

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+		}
+	},
+	watch: {
+		value(newVal) {
+			this.inputVal = this.value
+		}
+	},
+	beforeCreate: function() {
+		this.GenRandomId = GenRandomId.toString()
+		GenRandomId += 1
```
Comment (georgehrke) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: Why did you do this?

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+</script>
+```
+
+</docs>
+
+<template>
+	<div class="section">
+		<p>
+			<input :id="id" type="checkbox" class="checkbox"
+				:checked="inputVal" :disabled="disabled" @change="$emit('input', $event.target.checked)">
```
Comment (georgehrke) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: You should outsource the `$emit('input', $event.target.checked)` call into a function

Then you can also `$emit('update:value', $event.target.checked)` to make it properly work with v-model.

Ideally you would also emit the same events as `ActionCheckbox` for consistencies reason:
https://github.com/nextcloud/nextcloud-vue/blob/master/src/components/ActionCheckbox/ActionCheckbox.vue#L111L137

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+			default: false
+		},
+		disabled: {
+			type: Boolean,
+			default: false
+		}
+	},
+	data() {
+		return {
+			inputVal: this.value
```
Comment (georgehrke) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: Why do you do this?

In the template, you can just use value instead of inputValue.

**Comment by georgehrke** [comment]
Also, you can't trigger it by clicking enter.

You can check here how to do it: https://github.com/nextcloud/nextcloud-vue/blob/master/src/components/ActionCheckbox/ActionCheckbox.vue#L44

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+		</p>
+	</div>
+</template>
+
+<script>
+import GenRandomId from 'Utils/GenRandomId'
+export default {
+	name: 'SettingsCheckbox',
+	props: {
+		label: {
```
Comment (georgehrke) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: Please add comments to your props so they show up as comments in the documentation :)

File: src/components/SettingsCheckbox/SettingsCheckbox.vue (hunk 8 above)
Comment (GretaD) [reply src/components/SettingsCheckbox/SettingsCheckbox.vue]: John told me so 🤷🏼‍♀️ that we should use `GenRandomId` instead of `uuid`. I will fix the string part.

Comment (skjnldsv) [reply src/components/SettingsCheckbox/SettingsCheckbox.vue]: @georgehrke you need a random unique id for the label/id combination, so we don't have to provide one every time.

@GretaD we should still have the option to provide one, so I would create a prop and set the default with GenRandomId :)

File: src/components/SettingsCheckbox/SettingsCheckbox.vue (hunk 9 above)
Comment (georgehrke) [reply src/components/SettingsCheckbox/SettingsCheckbox.vue]: This still needs to be resolved.

Otherwise this is good to go

**Comment by skjnldsv** [comment]
What is the status on this? :)

**Comment by GretaD** [comment]
> What is the status on this? :)

I have pushed all the changes as requested by reviewers, maybe a final review and then we merge is everything is good

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+  - You should have received a copy of the GNU Affero General Public License
+  - along with this program. If not, see <http://www.gnu.org/licenses/>.
+  -
+  -->
+<docs>
+## One checkbox with a hint
+
+```vue
+<template>
+		<settings-checkbox
```
Comment (skjnldsv) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: We do not use kebab case components in vue
```suggestion
		<SettingsCheckbox
```

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+				 */
+				this.$emit('uncheck')
+			}
+		}
+	}
+}
+</script>
+
+<style lang="scss" scoped>
+p {
```
Comment (skjnldsv) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: If we really want to be independent, the styling of the whole checkbox needs to not rely on the server css.
This needs to be perfectly good looking on the nc-vue documentation https://deploy-preview-649--nextcloud-vue-components.netlify.com/

Comment (skjnldsv) [reply src/components/SettingsCheckbox/SettingsCheckbox.vue]: This goes for all of your settings pr ;)

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+</docs>
+
+<template>
+	<input
+		:id="id"
+		type="checkbox"
+		class="checkbox"
+		:checked="value"
+		:disabled="disabled"
+		@keydown.enter.exact.prevent="checkInput" @change="onChange">
```
Comment (georgehrke) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: Please put `@change` into a new line as well

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+	<label :for="id">{{ label }}</label><br>
```
Comment (georgehrke) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]: No `<br>` please :) 

If you want to force a new line, wrap it in a div (or some other element with `display:block`)

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+		 */
+		checked: {
+			type: Boolean,
+			default: false
+		},
+		/**
+		 * value of the checkbox input
+		 */
+		value: {
+			type: [String, Number],
```
Comment (georgehrke) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]:
```suggestion
			type: Boolean,
```

Since it's a Checkbox, the value should be a Boolean :)

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+		type="checkbox"
+		class="checkbox"
+		:checked="value"
+		:disabled="disabled"
+		@keydown.enter.exact.prevent="checkInput" @change="onChange">
+	<label :for="id">{{ label }}</label><br>
+</template>
+
+<script>
+
```
Comment (georgehrke) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]:
```suggestion
import GenRandomId from 'Utils/GenRandomId'

```

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+export default {
+	name: 'SettingsCheckbox',
+	props: {
+		/**
+		 * id attribute of the checkbox element
+		 */
+		id: {
+			type: String,
+			default: 'settings-checkbox-' + this.GenRandomId
```
Comment (georgehrke) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue]:
```suggestion
			default: () => 'settings-checkbox-' + GenRandomId()
```

**Comment by georgehrke** [comment]
And please revert the changes you made to `package.json` and `package-lock.json`

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+<template>
+	<input
+		:id="id"
+		type="checkbox"
+		class="checkbox"
+		:checked="value"
+		:disabled="disabled"
+		@keydown.enter.exact.prevent="checkInput"
+		@change="onChange">
+	<label :for="id">{{ label }}</label>
```
Comment (skjnldsv) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue:45]: Not valid, as Eslint says: 'The template root requires exactly one element'

You can put the input inside the label if you want, this is valid and will solve your issue of one root element. 
https://youtu.be/g9Qff0b-lHk

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+		checked: {
+			type: Boolean,
+			default: false
+		},
+		/**
+		 * value of the checkbox input
+		 */
+		value: {
+			type: Boolean,
+			default: false
```
Comment (skjnldsv) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue:77]: Value is not a boolean, it's a string that serves as an identifier for the form
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input/checkbox#Value

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+		/**
+		 * disabled state of the checkbox element
+		 */
+		disabled: {
+			type: Boolean,
+			default: false
+		}
+	},
+	methods: {
+		checkInput(event) {
```
Comment (skjnldsv) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue:88]: You do not need this if the `for`/`id` pair is properly matching.
Also if you nest them it should be done automatically I think :)

File: src/components/SettingsCheckbox/index.js [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+import SettingsCheckbox from './SettingsCheckbox'
+
+export default SettingsCheckbox
+export { SettingsCheckbox }
```
Comment (skjnldsv) [inline src/components/SettingsCheckbox/index.js:25]: Not needed anymore
```suggestion
```

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+		},
+		onChange(event) {
+			/**
+			 * Emitted when the checkbox state is changed
+			 * @type {boolean}
+			 */
+			this.$emit('update:checked', this.$refs.checkbox.checked)
+			/**
+			 * Emitted when the checkbox state is changed
+			 * @type {Event}
```
Comment (skjnldsv) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue:100]: Nice documentation! :+1:

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+  -
+  -->
+<docs>
+## One settingscheckbox with a hint
+
+```vue
+<template>
+		<SettingsCheckbox
+			label="some text"
+			hint="some hint"
```
Comment (skjnldsv) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue:30]: What does hint means here? It's not in the props?

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+			v-model="checkedTarget"/>
+</template>
+```
+</docs>
+
+<template>
+	<input
+		:id="id"
+		type="checkbox"
+		class="checkbox"
```
Comment (skjnldsv) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue:40]: Please remove the class so we're completely independent from the server's design
```suggestion
```

File: src/components/SettingsCheckbox/SettingsCheckbox.vue (hunk 19 above)
Comment (skjnldsv) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue:45]: Maybe use a default slot instead, I will find it cleaner
```vue
<SettingsCheckbox>This is the label</SettingsCheckbox>
```

File: src/components/SettingsCheckbox/SettingsCheckbox.vue [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+		checkInput(event) {
+			// by clicking we also trigger the change event
+			this.$refs.label.click()
+		},
+		onChange(event) {
+			/**
+			 * Emitted when the checkbox state is changed
+			 * @type {boolean}
+			 */
+			this.$emit('update:checked', this.$refs.checkbox.checked)
```
Comment (skjnldsv) [inline src/components/SettingsCheckbox/SettingsCheckbox.vue:97]: Another thing, we should also be able to use `v-model` here, so we should also emit an input with the checked status :thinking: 
https://vuejs.org/v2/guide/forms.html#Checkbox
https://stackoverflow.com/a/46262758/3885878
https://forum.vuejs.org/t/custom-checkbox-component/1120/7

File: src/components/SettingsCheckbox/SettingsCheckbox.vue (hunk 20 above)
Comment (skjnldsv) [reply src/components/SettingsCheckbox/SettingsCheckbox.vue:77]: Nonetheless, it might be used by v-model, so you will have to check.
If so, we'll have to find a common ground between using :checked and :value
https://vuejs.org/v2/guide/forms.html#Basic-Usage

@juliushaertl @georgehrke , any thoughts?

Comment (GretaD) [reply src/components/SettingsCheckbox/SettingsCheckbox.vue:77]: it was sting at my initial patch, but Georg proposed boolean: https://github.com/nextcloud/nextcloud-vue/pull/649#discussion_r385111608

**Comment by skjnldsv** [comment]
Great work @GretaD :)

We should maybe remove the SettingCheckbow component, now, no?
Or make it use the Checkbox one?

**Comment by GretaD** [comment]
> Great work @GretaD :)
> 
> We should maybe remove the SettingCheckbow component, now, no?
> Or make it use the Checkbox one?

Yes we should remove it, i thought i did :goat:

File: src/components/Checkbox/Checkbox.vue [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+		checkInput(event) {
+			// by clicking we also trigger the change event
+			this.$refs.label.click()
+		},
+		onChange(event) {
+			/**
+			 * Emitted when the checkbox state is changed
+			 * @type {boolean}
+			 */
+			this.$emit('update:checked', this.$refs.checkbox.checked)
```
Comment (georgehrke) [inline src/components/Checkbox/Checkbox.vue:89]: If you want to support `v-model`, like you gave an example of above, you will have to emit an event called `update:value`.
```suggestion
			this.$emit('update:value', this.$refs.checkbox.checked)
```

Comment (skjnldsv) [reply src/components/Checkbox/Checkbox.vue:89]: `update:value` is not for v-model. But for the sync modifier.
v-model will rely on change/input depending on the elemtn type :wink:

Comment (georgehrke) [reply src/components/Checkbox/Checkbox.vue:89]: True, my bad, sorry. But since there is no prop called `checked` it should be renamed anyway

File: src/components/Checkbox/Checkbox.vue [hunk 28]
```diff
@@ -0,0 +1,169 @@
...
+import GenRandomId from '../../utils/GenRandomId'
+export default {
+	name: 'Checkbox',
+	props: {
+		/**
+		 * id attribute of the checkbox element
+		 */
+		id: {
+			type: String,
+			default: () => 'checkbox-' + GenRandomId,
```
Comment (skjnldsv) [inline src/components/Checkbox/Checkbox.vue:57]: `GenRandomId` is a function

```suggestion
			default: 'checkbox-' + GenRandomId(5),
```

File: src/components/Checkbox/Checkbox.vue [hunk 29]
```diff
@@ -0,0 +1,169 @@
...
+		label: {
+			type: String,
+			required: true,
+		},
+		/**
+		 * value of the checkbox input
+		 */
+		value: {
+			type: Boolean,
+			default: false,
```
Comment (skjnldsv) [inline src/components/Checkbox/Checkbox.vue:69]:
```suggestion
			required: true,
```

Or am I missing something? :)

**Comment by GretaD** [comment]
Thank you Barth for the review. I will change them asap, but can you please have a general review and decide what you think is fine and what it is not, so i can change everything in one push. You reviewed this PR many times, and every time you found something new, which is fine, but why not adding all the changes in the first review. Let me know if you're done so i can push the changes. Thanks again.

File: src/components/Checkbox/Checkbox.vue [hunk 30]
```diff
@@ -0,0 +1,169 @@
...
+</docs>
+
+<template>
+	<label>
+		<input
+			:id="id"
+			type="checkbox"
+			:checked="value"
+			:disabled="disabled"
+			@keydown.enter.exact.prevent="checkInput"
```
[the remark this answers is not in the document]
Comment (jancborchardt) [reply src/components/Checkbox/Checkbox.vue:41]: Fully agree with @skjnldsv, we should make sure to observe the browser standards. :)

**Comment by skjnldsv** [comment]
> You reviewed this PR many times, and every time you found something new, which is fine, but why not adding all the changes in the first review. Let me know if you're done so i can push the changes.

Sorry :see_no_evil: 
Yes, it should be fine after the last changes I think :)
You're getting close!! :muscle:

File: src/components/Checkbox/Checkbox.vue [hunk 31]
```diff
@@ -0,0 +1,167 @@
...
+		.checkbox__label {
+			cursor: pointer;
+			display: inline-block;
+		}
+	}
+	&:not(.checkbox--disabled):hover,
+	&:not(.checkbox--disabled):focus {
+		.checkbox__label {
+			opacity: $opacity_full;
+		}
```
Comment (skjnldsv) [inline src/components/Checkbox/Checkbox.vue:163]: Did you lose some code while rebasing or something?
This seems unrelated now?

File: src/components/Checkbox/Checkbox.vue [hunk 32]
```diff
@@ -0,0 +1,167 @@
...
+  -->
+<docs>
+## One checkbox
+
+```vue
+<template>
+		<Checkbox
+			label="some text"
+			v-model="checkedTarget"/>
+</template>
```
Comment (skjnldsv) [inline src/components/Checkbox/Checkbox.vue:33]: checkedTarget is undefined

```suggestion
<template>
		<Checkbox
			label="some text"
			v-model="checkedTarget"/>
</template>
<script>
 export default {
 	data() {
 		return {
 			checkedTarget: false
 		}
 	},
 }
</script>
```

**Review by skjnldsv — CHANGES_REQUESTED** [review]
Design is off and not fitting the nextcloud checkbox design 

This pr:
Capture d’écran_2020-10-30_13-12-36


Wanted:
Capture d’écran_2020-10-30_13-13-25

**Comment by skjnldsv** [comment]
> and every time you found something new, which is fine, but why not adding all the changes in the first review

I think this is because this pr changed so much I find new stuff every time.
Please start over and check this is exactly done and working as it is supposed to.
I don't understand what happened to be honest :confused:

**Comment by GretaD** [comment]
@skjnldsv what do you think about closing this one and I start a new one when i have the time to start and finish it on time. This is getting very old. And i can add your suggestions on the new one.

**Comment by GretaD** [comment]
closing to open a new one
